### PR TITLE
Python Command

### PR DESCRIPTION
**Pull request by cartertinney**
Raw Paho MQTT implementation of command with minimalist infrastructure

### CODE REVIEW

File: scenarios/command/python/mobile_app.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+    msg_prop.UserProperty = ("When", str(datetime.datetime.utcnow()))
+    msg_prop.UserProperty = ("RequestedFrom", client_id)
+    msg_prop.CorrelationData = correlation_id
+    print("Sending Unlock Request")
+    message_info = mqtt_client.publish(topic, payload, qos=1, properties=msg_prop)
+    message_info.wait_for_publish(timeout=10)
+    response_future = request_ledger.get_response_future(correlation_id)
+    print("Waiting for Unlock Response")
+    response_msg = response_future.result(timeout=10)
+    print("Response: {}".format(response_msg.Succeed))
```
Comment (olivakar) [inline scenarios/command/python/mobile_app.py:135]: DO
`print("Response: {}".format(response_msg.properties.UserProperty[0]))`

File: scenarios/command/python/vehicle_03.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    # # In Paho CB thread.
+    with connected_cond:
+        connected_prop = False
+        connected_cond.notify_all()
+
+def send_unlock_response(mqtt_client, correlation_id):
+    topic = RESPONSE_TOPIC_PATTERN.format(targetClientId="vehicle03", commandName="unlock")
+    payload = "placeholder"
+    msg_prop = Properties(PacketTypes.PUBLISH)
+    msg_prop.UserProperty = ("Succeed", True)
```
Comment (olivakar) [inline scenarios/command/python/vehicle_03.py:67]: Do
`msg_prop.UserProperty = ("Succeed", "True")`

**Review by olivakar — APPROVED** [review]

**Review by olivakar — APPROVED** [review]

File: scenarios/command/python/mobile_app.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+def on_subscribe(client, _userdata, mid, _reason_codes, _properties):
+    global subscribed_prop
+    print(f"Subscribe for message id {mid} acknowledged by MQTT broker")
+    # # In Paho CB thread.
+    with subscribed_cond:
+        subscribed_prop = True
+        subscribed_cond.notify_all()
+
+def on_unlock_response(_client, _userdata, message):
+    #print(f"Received message on topic {message.topic} with payload {message.payload}")
```
Comment (vaavva) [inline scenarios/command/python/mobile_app.py:56]: Should this get un-commented?

File: scenarios/command/python/mobile_app.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        mqtt_client.tls_set_context(context)
+    return mqtt_client
+
+def send_unlock_command(mqtt_client, client_id):
+    topic = REQUEST_TOPIC_PATTERN.format(targetClientId="vehicle03", commandName="unlock")
+    payload = "placeholder"
+    correlation_id = str(uuid.uuid4()).encode()
+    msg_prop = Properties(PacketTypes.PUBLISH)
+    msg_prop.UserProperty = ("When", str(datetime.datetime.utcnow()))
```
Comment (vaavva) [inline scenarios/command/python/mobile_app.py:124]: I think When and RequestedFrom should be in the payload, not user properties

File: scenarios/command/python/mobile_app.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    msg_prop.UserProperty = ("RequestedFrom", client_id)
+    msg_prop.CorrelationData = correlation_id
```
Comment (vaavva) [inline scenarios/command/python/mobile_app.py:126]: Are you sending the response topic on the unlock command?

File: scenarios/command/python/vehicle_03.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import uuid
+import ssl
+import threading
+import paho.mqtt.client as mqtt
+from paho.mqtt.properties import Properties
+from paho.mqtt.packettypes import PacketTypes
+from concurrent.futures import ThreadPoolExecutor
+
+REQUEST_TOPIC_PATTERN = "vehicles/{targetClientId}/command/{commandName}/request"
+RESPONSE_TOPIC_PATTERN = "vehicles/{targetClientId}/command/{commandName}/response"
```
Comment (vaavva) [inline scenarios/command/python/vehicle_03.py:19]: We should be getting this from the unlock request

File: scenarios/command/python/vehicle_03.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    # # In Paho CB thread.
+    with connected_cond:
+        connected_prop = False
+        connected_cond.notify_all()
+
+def send_unlock_response(mqtt_client, correlation_id):
+    topic = RESPONSE_TOPIC_PATTERN.format(targetClientId="vehicle03", commandName="unlock")
+    payload = "placeholder"
+    msg_prop = Properties(PacketTypes.PUBLISH)
+    msg_prop.UserProperty = ("Succeed", "True")
```
Comment (vaavva) [inline scenarios/command/python/vehicle_03.py:67]: Should be in payload, not user property

Comment (rido-min) [reply scenarios/command/python/vehicle_03.py:67]: we are using properties to indicate status.. what needs to be in the payload?

Comment (vaavva) [reply scenarios/command/python/vehicle_03.py:67]: In the C# sample, Succeed: True is in the payload, and Status: 200 is in a user property - it seems reasonable to me if we don't need to have anything in the payload on the response though

File: scenarios/command/python/mobile_app.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    response_future = request_ledger.get_response_future(correlation_id)
+    print("Waiting for Unlock Response")
+    response_msg = response_future.result(timeout=10)
+    print("Response: {}".format(response_msg.properties.UserProperty[0]))
+
+
+def main():
+    connection_settings = cs.get_connection_settings(args.env_file)
+    if not connection_settings["MQTT_CLEAN_SESSION"]:
+        raise ValueError("This sample does not support connecting with existing sessions")
```
[the remark this answers is not in the document]
Comment (vaavva) [reply scenarios/command/python/mobile_app.py:139]: Right now none of the samples properly support clean session = false, so I think we should add this error messaging to all of them until we support it

**Comment by cartertinney** [comment]
> It's ok to start, but I was expecting to see the code splitted into three layers:
> 
> * Common. Helper funcions using in all samples, such as connecting with settings
> * Binders. Encapsulate topics and serialization formats
> * Clients. Just using the binders without any mqtt concepts

This kind of division is not very viable within the structure of the samples we're discussing here, which is why we have elected to implement samples directly with raw Paho instead. This is due to differences in how the MQTT client works in Python, as well as various language features that make mimicking the C# samples highly difficult and convoluted. I can explain the technical details in depth if you wish, please just reach out.

However, I can assure you that the pattern you desire **will** be reflected in the real SDK lite implementation, which we have already begun work on.

File: scenarios/command/README.md [hunk 9]
```diff
@@ -201,3 +201,14 @@ To run the dotnet sample execute each line below in a different shell/terminal.
 ```bash
  dotnet/command_consumer/bin/Debug/net7.0/command_consumer --envFile=mobile-app.env
 ```
+
+### python
```
Comment (vaavva) [inline scenarios/command/README.md:205]: I think this is required as well to be able to run this:
https://github.com/Azure-Samples/MqttApplicationSamples/blob/main/scenarios/telemetry/README.md?plain=1#L260-L267

**Pull request merged**